_drafts/c-struct-padding-initialization.md

```diff
@@ -0,0 +1,460 @@
+| 2, explicit | `2`                | zero      | **unset** |
+| 2, explicit | `3`                | zero      | **unset** |
+| 2, explicit | `s`                | **unset** | **unset** |
+
+### Strategy 3, `{ 0 }`
+
+| Strategy   | Optimization Level | Clang 13  | GCC 11 |
+| ---------- | ------------------ | --------- | ------ |
+| 3, `{ 0 }` | `0`                | zero      | zero   |
+| 3, `{ 0 }` | `1`                | **unset** | zero   |
+| 3, `{ 0 }` | `2`                | zero      | zero   |
+| 3, `{ 0 }` | `3`                | zero      | zero   |
+| 3, `{ 0 }` | `s`                | zero      | zero   |
+
+### Strategy 4, `{ }` GCC extension
+
+| Strategy | Optimization Level | Clang 13  | GCC 11 |
+| -------- | ------------------ | --------- | ------ |
+| 4, `{ }` | `0`                | zero      | zero   |
+| 4, `{ }` | `1`                | **unset** | zero   |
+| 4, `{ }` | `2`                | zero      | zero   |
+| 4, `{ }` | `3`                | zero      | zero   |
+| 4, `{ }` | `s`                | **unset** | zero   |
+
+The main point is that it's not particulary consistent across compilers and
+optimization levels 😱!
+
+> You can find the example application used to generate the above data here on
+> [Github](https://github.com/memfault/interrupt/tree/master/example/c-struct-padding-initialization/).
+
+> I've also uploaded it to the wonderful Compiler Explorer if you want to take a
+> look and quickly play around: <https://godbolt.org/z/b985G4ejT>
+
+## Best Practice?
+
+It's tricky to recommend a one-size-fits-all option here, because different
+software will have different constraints. However, some general purpose advice
+follows.
+
+### Avoid relying on structure layout
+
+The simplest option to avoid padding issues is to avoid the padding fields
+altogether:
+
+> Access data directly via each member, do not alias structures or use `memcmp`
+> etc.
+
+This only works if the padding data can be safely ignored in all use cases for
+the data structures in question.
+
+Another approach is to avoid structure holes entirely! The `pahole` tool could
+be used to detect any structures with padding bits (for example, as a linter
+pass on the generated binary!), and they can be corrected by either reordering
+structure members to eliminate padding, or adding `uint8_t unused_[n]` fields to
+explicitly address the holes. See also [The Lost Art of Structure
+Packing](http://www.catb.org/esr/structure-packing/) linked previously.
+
+### Use `memset` to zero-initialized padding bits
+
+`memset` reliably sets the entire memory space of an object, including the
+padding bits.
+
+Must be manually done, though, so can be error-prone.
+
+Be sure to set the size argument directly from the object in question:
+
+```c
+// error prone! if the type of a changes, we might get unexpected results
+memset(&a, 0, sizeof(struct foo));
+
+// much better
+memset(&a, 0, sizeof(a));
+```
+
+### Last resort, `__attribute__((packed))`
+
```

Comment:
I do think this should ALWAYS be used for data that may be read outside the application currently running whether that be another app or a future update of the same app where the data is reloaded from flash.

Reply:
good call, i was kinda stretching for anti-arguments but it really is best practice!